### DISCUSSION TITLE
Multilabel feature selection

### DISCUSSION THREAD

**Discussion by pr4deepr** [blue-yonder/tsfresh#1041] [Q&A]
Hi
   I have a timeseries data where the classes are Control, drug 1, drug 2 and drug 3. I have run the default feature extraction and now would like to select relevant features.
I've been following this example:
https://github.com/blue-yonder/tsfresh/blob/main/notebooks/04%20Multiclass%20Selection%20Example.ipynb

I would like to select features that are relevant and quite different between control vs drug 1, control vs drug 2, control vs drug 3. I'm also interested in features different between drug treatments. How do I do approach this?

Thanks
Pradeep

**Comment by pr4deepr** [comment]
I think I may have an approach to this
In the **Multiclass feature selection** for the python notebook above, I can use set difference method instead of union. 
Only difference is that I store the relevant features for each condition in a dictionary

if labels are "control", "drug1", "drug2", then

```
relevant_feature_dict= {}

for label in y.unique():
    y_train_binary = y_train == label
    X_train_filtered = select_features(X_train, y_train_binary)
    print("Number of relevant features for class {}: {}/{}".format(label, X_train_filtered.shape[1], X_train.shape[1]))
    relevant_feature_dict[label] = X_filtered_mask.columns

#to get relevant features in drug1 which isn't in control
drug1_ctrl_features = set(relevant_feature_dict['drug1']).difference(set(relevant_feature_dict['control']))
```

Still experimenting with these. happy to hear any feedback..

**Reply by nils-braun** [reply]
I think that makes sense! 
One thing you might also want to do is select the features for drug1 only on the sample drug1 vs control (which does not contain drug2 and drug3) - depending on your use case that might (or might not) make more sense. With that, you already have the features which differentiate best between drug1 and control (without a multiclass problem).

Did you already get reasonable results with your approach?

**Reply by pr4deepr** [reply]
Thanks @nils-braun 
Appreciate the answer. Thats a really good point. I'll try that approach as well. 

Currently, the way I've went about it is use a tree-based approach to train a model and investigate the feature importance within the model.. Its still a work in progress and I'll update the thread/share my approach once I've gotten something solid!

Although, the area where I'm having difficulty is mostly understanding the meaning of the features. 
For example, 
* feature1_agg_linear_trend__attr_"rvalue"__chunk_len_5__f_agg_"var" seems to be a top hit... 
Another one: 
feature2__cwt_coeficcients__coeff_3_w_20__widths_(2,5,10,20)

The features and calculations are well documented, but its still a bit confusing..
Any advice for this?

**Reply by nils-braun** [reply]
"Funnily" enough, I am not an expert in most of the feature calculators myself. I am looping in @kempa-liehr - I remember he had some nice visualizations for some of the feature calculators to help understand what they do?
